ruma-federation-api/src/identification/identifier_bound/v1.rs

```diff
@@ -0,0 +1,80 @@
+//! [PUT /_matrix/federation/v1/3pid/onbind](https://matrix.org/docs/spec/server_server/r0.1.4#put-matrix-federation-v1-3pid-onbind)
+
+use ruma_api::ruma_api;
+use ruma_common::thirdparty::Medium;
+use ruma_identifiers::{RoomId, ServerKeyId, ServerNameBox, UserId};
+use serde::{Deserialize, Serialize};
+
+use std::collections::BTreeMap;
+
+ruma_api! {
+    metadata: {
+        description: "Used by identity servers to notify the homeserver that one of its users has bound a third party identifier successfully",
+        method: PUT,
+        name: "identifier_bound",
+        path: "/_matrix/federation/v1/3pid/onbind",
+        rate_limited: false,
+        requires_authentication: false,
+    }
+
+    #[non_exhaustive]
+    request: {
+        /// The type of third party identifier. Currently only "email" is a possible value.
+        pub medium: Medium,
+
+        /// The third party identifier itself. For example, an email address.
+        pub address: &'a str,
+
+        /// The user that is now bound to the third party identifier.
+        pub mxid: UserId,
+
+        /// A list of pending invites that the third party identifier has received.
+        pub invites: &'a [ThirdPartyInvite],
+    }
+
+    #[non_exhaustive]
+    response: {}
+}
+
+impl<'a> Request<'a> {
+    /// Creates a new `Request` with the given address, matrix id and third party invites.
+    pub fn new(address: &'a str, mxid: UserId, invites: &'a [ThirdPartyInvite]) -> Self {
+        Self { medium: Medium::Email, address, mxid, invites }
+    }
+}
+
+/// A pending invite the third party identifier has received.
+#[derive(Debug, Clone, Deserialize, Serialize)]
+#[non_exhaustive]
```

Comment:
The non-exhaustive attributes are now feature-gated.

```suggestion
#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
```